waiter/src/waiter/process_request.clj

```diff
@@ -405,10 +405,21 @@
           (async/close! body)
           (async/close! request-state-chan))))))
 
-(defn- track-response-status-metrics
-  [service-id {:strs [metric-group]} status]
-  (counters/inc! (metrics/service-counter service-id "response-status" (str status)))
-  (statsd/inc! metric-group (str "response_status_" status)))
+(defn wrap-response-status-metrics
+  "Wraps a handler and updates service metrics based upon the result."
+  [handler]
+  (fn wrap-response-status-metrics-fn
+    [{:keys [descriptor] :as request}]
+    (let [{:keys [service-id] {:strs [metric-group]} :service-description} descriptor
+          response (handler request)
+          update (fn [{:keys [status] :as response}]
+                   (counters/inc! (metrics/service-counter service-id "response-status" (str status)))
+                   (statsd/inc! metric-group (str "response_status_" status))
+                   response)]
+      (if (au/chan? response)
```

Comment:
We should have a helper function for the if-else pattern.

Reply:
I'm not sure this rises to the level of needed a helper.

Reply:
I'll create it in #231 and then rebase.

Reply:
Okay